consul/src/main/java/com/linecorp/armeria/internal/consul/AgentServiceClient.java

```diff
@@ -0,0 +1,220 @@
+     * Builds an AgentServiceClient with a ConsulClient.
+     */
+    public static AgentServiceClient of(ConsulClient consulClient) {
+        return new AgentServiceClient(consulClient);
+    }
+
+    private AgentServiceClient(ConsulClient client) {
+        this.client = client;
+    }
+
+    /**
+     * Gets a list of services from consul agent.
+     */
+    public CompletableFuture<Map<String, Service>> services(
+            @Nullable String serviceName, ConsulParameters parameters) {
+        String serviceFilter = "";
+        if (serviceName != null) {
+            serviceFilter = "filter=Service==" + serviceName + '&';
+        }
+        return client.consulWebClient()
+                     .get("/agent/services?" + serviceFilter + parameters.toParameterString())
+                     .aggregate()
+                     .thenApply(response -> {
+                         final ObjectMapper mapper = new ObjectMapper();
+                         try {
+                             return mapper.readValue(
+                                     response.content().toStringUtf8(),
+                                     mapper.getTypeFactory().constructMapType(
+                                             Map.class, String.class, Service.class));
+                         } catch (JsonProcessingException e) {
+                             throw new RuntimeException(e);
+                         }
+                     });
+    }
+
+    /**
+     * Registers a service.
+     */
+    public CompletableFuture<String> register(String serviceName, String address, int port,
+                                              ConsulParameters parameters) {
+        return register(serviceName, address, port, null, parameters);
+    }
+
+    /**
+     * Registers a service with a health checking endpoint.
+     *
+     * @return returns a generated service ID.
+     */
+    public CompletableFuture<String> register(String serviceName, String address, int port,
+                                              @Nullable Check check, ConsulParameters parameters) {
+        final String serviceId = UUID.randomUUID().toString();
+        return register(serviceId, serviceName, address, port, check, parameters);
+    }
+
+    /**
+     * Registers a service into the consul agent.
+     */
+    public CompletableFuture<String> register(String serviceId, String serviceName, String address, int port,
+                                            @Nullable Check check, ConsulParameters parameters) {
+        final Service service = new Service();
+        service.id = serviceId;
+        service.name = serviceName;
+        service.address = address;
+        service.port = port;
+        if (check != null) {
+            service.check = check;
+        }
+
+        final String jsonBody;
+        try {
+            jsonBody = client.getObjectMapper().writerWithDefaultPrettyPrinter().writeValueAsString(service);
+        } catch (JsonProcessingException e) {
```

Comment:
Just throw?

Reply:
Could you suggest treating the exception in the lambda expression also? (e.g.: around line 70th in `AgentServiceClient`)

Reply:
That's a good question. As you know Java cannot throw a checked exception in lambda expression so I suggested wrapping the checked exception with the runtime exception.
However, this is executed in the current method, not lambda so that I think we don't need to wrap `JsonProcessingException` with the runtime exception. We can just add `throws JsonProcessingException` to the method signature.

Reply:
Ok, I see.

Reply:
Please see https://github.com/line/armeria/pull/2192#discussion_r391994708 😀